File: Examples/intel_examples/examples_core_c/transc/source/comatadd.c
/*******************************************************************************
* Copyright 2010-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!  mkl_comatadd - out-of-place transposition routine,
!  Example Program Text ( C Interface )
!******************************************************************************/
#include <mkl_trans.h>
#include "common_func.h"

int main(int argc, char *argv[])
{ 
  size_t n=4, m=3; /* rows, cols of source matrix */
  MKL_Complex8 alpha;
  MKL_Complex8 beta;
  MKL_Complex8 a[]={
    { 1.,  2.},  { 3.,  4.}, { 5.,  6.},
    { 7.,  8.},  { 9., 10.}, {11., 12.},
    {13., 14.},  {15., 16.}, {17., 18.},
    {25., 26.},  {27., 28.}, {29., 30.}
  }; /* source matrix */
  MKL_Complex8 b[]={
    { 1.1,  2.1}, { 3.2,  4.2}, { 5.3,  6.3},
    { 7.1,  8.1}, { 9.2, 10.2}, {11.3, 12.3},
    {13.1, 14.1}, {15.2, 16.2}, {17.3, 18.3},
    {25.1, 26.1}, {27.2, 28.2}, {29.3, 30.3}
  }; /* source matrix   */
  MKL_Complex8 dst[9];/* destination matrix */
  alpha.real = 1.;
  alpha.imag = 0.;
  beta.real = 1.;
  beta.imag = 0.;

  printf("\nExample of using mkl_comatadd transposition\n");
  printf("INPUT DATA:\nSource matrix A:\n");
  print_matrix(n, m, 'c', a);

  printf("Source matrix B:\n");
  print_matrix(n, m, 'c', b);

  /* 
  **  Addition of transposed submatrix(3,3) a and unchanged submatrix(3,3) b 
  */
  mkl_comatadd('R'    /* row-major ordering */, 
               'T'    /* A will be transposed */,
               'N'    /* no changes to B */, 
                3     /* rows */, 
                3     /* cols */, 
                alpha /* alpha */, 
                a     /* source matrix */, 
                3     /* lda */, 
                beta  /* beta */, 
                b     /* source matrix */, 
                3     /* ldb */, 
                dst   /* destination matrix */, 
                3     /* ldc */); 
  /*
  **  New matrix: c =  { 
  **    2.1, 4.1,     10.2, 12.2,    18.3, 20.3,
  **    10.1, 12.1,    18.2, 20.2,    26.3, 28.3,
  **    18.1, 20.1,    26.2, 28.2,    34.3, 36.3
  **  }
  */
  printf("OUTPUT DATA:\nDestination matrix - addition of transposed submatrix(3,3) A and submatrix B:\n");  
  print_matrix(3, 3, 'c', dst);
  /*
  **  Addition of transposed submatrix(3,3) a and conjugate transposed submatrix(3,3) b 
  */
  mkl_comatadd('R'    /* row-major ordering */, 
               'T'    /* A will be transposed */,
               'C'    /* B will be conjugate transposed */, 
                3     /* rows */, 
                3     /* cols */, 
                alpha /* alpha */, 
                a     /* source matrix */, 
                3     /* lda */, 
                beta  /* beta */, 
                b     /* source matrix */, 
                3     /* ldb */, 
                dst   /* destination matrix */, 
                3     /* ldc */);  
  /*  New matrix: c = {
  **    2.1, -0.1,  14.1, -0.1,  26.1, -0.1,
  **    6.2, -0.2,  18.2, -0.2,  30.2, -0.2,
  **    10.3, -0.3, 22.3, -0.3,  34.3, -0.3   
  **  }
  */
  printf("Destination matrix - Addition of transposed submatrix(3,3) A and conjugate transposed submatrix(3,3) B:\n");  
  print_matrix(3, 3, 'c', dst);
  
  return 0;
}

File: Examples/intel_examples/examples_core_f/transf/source/comatadd.f90
!===============================================================================
! Copyright 2010-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!      C O M A T A D D  Example Program Text
!*******************************************************************************
    program  COMATADD_MAIN
    include 'mkl_trans.fi' 

    integer :: i, j
    complex :: alpha = (1.0,0.0), beta = (1.0,0.0)
!   source matrix A
    complex,dimension(3,5),parameter :: A=RESHAPE(&
                                         &(/(1.0,1.0), (5.0,5.0), ( 9.0, 9.0),& 
                                           &(2.0,2.0), (6.0,6.0), (10.0,10.0),& 
                                           &(3.0,3.0), (7.0,7.0), (11.0,11.0),&
                                           &(4.0,4.0), (8.0,8.0), (12.0,12.0),&
                                           &(0.0,0.0), (0.0,0.0), ( 0.0, 0.0)/),SHAPE(A))
!   source matrix B
    complex,dimension(4,2),parameter :: B=RESHAPE(&
                                         &(/(1.0,0.0), (4.0,0.0), (7.0,0.0), (10.0,0.0),&  
                                           &(2.0,0.0), (5.0,0.0), (8.0,0.0), (11.0,0.0)/),SHAPE(B))
    complex,dimension(3,2) :: C ! destination matrix 1 
    complex,dimension(2,3) :: D ! destination matrix 2

!   Executable Statements
    print*, ' Example of using mkl_comatadd transposition '
    print*, ' INPUT DATA: '
    print*, ' Source matrix A: '    
!   print source matrix           
    do i= 1, 3
          print 102, ( A(i, j), j=1, 5)
    enddo
    print*, ' Source matrix B: '    
!   print source matrix           
    do i= 1, 4
          print 102, ( B(i, j), j=1, 2)
    enddo
! 
!   Addition of transposed submatrix(3, 3) A and 
!   unchanged submatrix(3, 3) B     
!
    call MKL_COMATADD( 'C',    & ! column-major ordering
                     & 'C',    & ! A is transposed
                     & 'R',    & ! B is unchanged
                     &  3,     & ! rows of destination matrix
                     &  2,     & ! cols of destination matrix
                     &  alpha, & ! alpha
                     &  A,     & ! source matrix
                     &  3,     & ! lda
                     &  beta,  & ! beta
                     &  B,     & ! source matrix
                     &  4,     & ! ldb
                     &  C,     & ! destination matrix
                     &  3 )      ! ldc
!   New matrix: c = ( 
!     (2.,-1.),   (7., -5.),  
!     (6.,-2.),  (11., -6.),
!    (10.,-3.),  (15., -7.)
!   )  
    print*, 'OUTPUT DATA: '
    print*, 'Destination matrix: '      
    print*, 'Addition of transposed submatrix(3,3) of A &
            & and submatrix(3,3) of B:'
!   print destination matrix      
    do i= 1, 3
          print 102,  ( C(i, j), j = 1, 2)
    enddo
! 
!   Addition of transposed submatrix(3, 3) A and 
!   transposed submatrix(3, 3) B     
!
    call MKL_COMATADD( 'C',    & ! column-major ordering
                     & 'T',    & ! A is transposed
                     & 'T',    & ! B is transposed
                     &  2,     & ! rows of destination matrix
                     &  3,     & ! cols of destination matrix
                     &  alpha, & ! alpha
                     &  A,     & ! source matrix
                     &  3,     & ! lda
                     &  beta,  & ! beta
                     &  B,     & ! source matrix
                     &  4,     & ! ldb
                     &  D,     & ! destination matrix
                     &  2 )      ! ldd
!   New matrix: D = { 
!     (2., 1.),  (9.,  5.), (16.,  9.),
!     (4., 2.), (11.,  6.), (18., 10.)
!   )         
    print*, 'Destination matrix: '      
    print*, 'Addition of transposed submatrices(3,3) of A and B: '
!   print destination matrix      
    do i= 1, 2
          print 102, (D(i, j), j = 1,3)
    enddo
     
102 format(9x,10(f6.2,',',f6.2,'i',3x))
    
    stop
    end